qi-doc/scribblings/macros.scrbl

```diff
@@ -0,0 +1,153 @@
+]
+
+@subsection{Racket Version Compatibility}
+
+ As binding spaces were added to Racket in version 8.3, older versions of Racket will not be able to use the macros described here, but can still use the legacy @seclink["Language_Extension"]{@racket[qi:]-prefixed macros}.
+
+@section{Example Macros}
+
+When you consider that Racket's @seclink["classes" #:doc '(lib "scribblings/guide/guide.scrbl")]{class-based object system} for object-oriented programming is implemented with Racket macros in terms of the underlying @seclink["structures" #:doc '(lib "scribblings/reference/reference.scrbl")]{struct type} system, it gives you some idea of the extent to which macros enable the addition of new language features, both great and small. In this section we'll look at a few examples of what Qi macros can do.
+
+@subsection{Write Yourself a Maybe Monad for Great Good}
+
+In functional languages such as Haskell, a popular way to do (or rather avoid) exception handling is to use the Maybe monad. Qi doesn't include monads out of the box yet, but you could implement a version of the Maybe monad yourself by using macros. But first, let's quickly review why you might want to in the first place.
+
+Earlier, we @seclink["Overview" #:doc '(lib "qi/scribblings/qi.scrbl")]{drew a distinction} between two paradigms employed in programming languages: one organized around the flow of @emph{control} and another organized around the flow of @emph{data}. A way to manage possible errors in code along the lines of the former ("control") paradigm is to handle @emph{exceptions} that may occur at each stage, and take appropriate action -- for instance, abort the remainder of the computation. A second way to handle errors, more along the lines of the "flow of data" paradigm, is for the "failing" computation to simply produce a sentinel value that signifies an error, so that the sequence of operations does not actually fail but merely generates and propagates a value signifying failure. The trick is, how to do this in such a way that downstream computations are aware of the sentinel error value so that they don't attempt to perform computations on it that they might do on a "normal" value? This is where the Maybe monad comes in.
+
+Let's say that we want to thread values through a number of flows, and if any of those flows raises an exception, we'd like the entire flow to generate @emph{no values}. We can start by writing a macro that wraps any Qi flow with the exception handling logic to generate no values.
+
+@racketblock[
+(define-qi-syntax-rule (try flo)
+  (esc (λ args
+         (with-handlers [(exn? (☯ ⏚))]
+           (apply (☯ flo) args)))))
+]
+
+This flow escapes to Racket in order to wrap the flow with exception handling. Any exceptions raised by execution of the flow result in the enclosing flow simply generating no values.
+
+And now, we're ready to write our Maybe monad.
+
+@racketblock[
+(define-qi-syntax-rule (maybe~> flo ...)
```

Comment:
I thought you might enjoy that example :) You're right that there is no explicit `unit` and `bind` here, but I believe this does satisfy the monad laws!

My Haskell is rudimentary, but I think this actually implements the Kleisli composition operator [`>=>`](http://dev.stephendiehl.com/hask/#kleisli-category), in terms of which the laws are :

```
(f >=> g) >=> h ≡ f >=> (g >=> h)
return >=> f ≡ f
f >=> return ≡  f
```

The input to each stage of the pipeline is the "monadic" value expressed in terms of presence or absence (unlike in the case of Maybe where there is a reified absence value -- I think the Qi version is more elegant!). Now Kleisli composition composes functions like:
```
f : a -> Mb
```
In our case that's e.g. `(try add1)`, since this accepts a normal value and either produces or doesn't produce a value. And indeed, the composition we've defined here takes two such functions:

```
f: (try sqr)
g: (try add1)
```

and composes them as:
```
(~> (when live? f)
    (when live? g))
```
We can call this operation `mcomp` (same as `>=>`) and define it as:
```
(mcomp g f) = (~> (when live? f)
                  (when live? g))
```
This composed function accepts a normal value and either generates a value, or doesn't. That is, it has the signature `g . f :: a -> Mc`. And it satisfies associativity since:

```
(~> (when live? f)
    (~> (when live? g)
        (when live? h)))
```
is equal to:
```
(~> (~> (when live? f)
        (when live? g))
    (when live? h))
```
... since when a value is defined at every stage it reduces to normal function composition. And when it isn't, both cases result in none being produced.

The "return" function is kind of interesting. For Haskell's Maybe it's "just" so that `3 -> Just 3`, a _distinct_ value. But here, the monadic value is simply existence or voidness. So "return" is somewhat metaphysical since it doesn't map an existing value to a monadic value. Rather, the scheme as a whole _lends_ the existence or absence of a value a special interpretation. As such, we can simply use `_` as return.

Then:
`return >=> f` is `(mcomp _ f)` which is
```
(~> (~> (when live? _)
        (when live? f)))
```
which reduces to:
```
(~> (~> (when live? f))) -> (~> (when live? f)) -> (when live? f)
```
And `f >=> return` is `(mcomp f _)` which is
```
(~> (~> (when live? f)
        (when live? _)))
```
which reduces to:
```
(~> (~> (when live? f))) -> (~> (when live? f)) -> (when live? f)
```
The requirement is that `f >=> return` as well as `return >=> f` should be `f`, but we have `(when live? f)` instead!

This seems problematic. We expect the result to be `f :: a -> b`, not `f' :: Ma -> b`

The resolution is in the nature of `return` -- in making the value "monadic," it does not add anything material to the underlying value, _only an interpretation_ of its presence or absence. Therefore, in all cases where there is a value present, i.e. in all cases where `f` is defined, `(when live? f)` is precisely equal to `f`. As it happens, this result is technically also defined on a larger domain than f itself, but as it produces no values in that case, it isn't a function in the usual sense. To the extent that we're comparing apples and apples, they are the same.

And so, we have a Maybe monad 😄 🥁 

(Btw on this subject, I was trying to wrap my head around monad transformers in haskell which appear to be a way to layer more than one monad on top of the ground computation. It looks pretty complicated in Haskell, but I think it could be really simple in Qi, although the implementation would be simplest if we had bindings. So, this could be a great case where bindings would make things clean!).

Reply:
It does look a lot more like a monad when you define `mcomp`! The section could feel a lot more Monad-y if you defined all the bits of the monad and then compose them together to define `maybe~>`. Without the pieces like `mcomp` the connection doesn't click for me.

Given the type for composition

```
(>=>) :: Monad m => (a -> m b) -> (b -> m c) -> a -> m c
```

it seems like the result of composition should expect a base value as argument, not a monadic value. So wouldn't it be

```
(mcomp a b) = (~> a (when live? b))
```

instead? I think that fixes the problem with `f >=> return`.

Another bit that seems off to me is that once you get to an error, the control flow keeps passing along that error through the remaining `(when live? (try flo))` elements instead of just directly returning `⏚`. My impression is that usually the Maybe monad implementation just avoids calling the pieces later in the composition. Perhaps this could work:

```
(define-qi-syntax-parser maybe~>
  [(_ flo)
    #'(try flo)]
  [(_ flo1 flo ...)
   #'(mcomp (try flo1) (maybe~> flo ...))])
```